File: demo-wiremock/src/test/java/io/fries/wiremock/demo/wiremock/cucumber/wiremock/record/RecordConfiguration.java
package io.fries.wiremock.demo.wiremock.cucumber.wiremock.record;

import com.github.tomakehurst.wiremock.WireMockServer;
import com.github.tomakehurst.wiremock.extension.responsetemplating.ResponseTemplateTransformer;
import com.github.tomakehurst.wiremock.recording.RecordSpec;
import io.fries.wiremock.demo.wiremock.cucumber.wiremock.MockServer;
import io.fries.wiremock.demo.wiremock.cucumber.wiremock.MockServers;
import io.fries.wiremock.demo.wiremock.cucumber.wiremock.WireMockProperties;
import io.fries.wiremock.demo.wiremock.cucumber.wiremock.WireMockProperties.WireMockServerProperties;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.Profile;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;

import static com.github.tomakehurst.wiremock.client.WireMock.recordSpec;
import static com.github.tomakehurst.wiremock.core.WireMockConfiguration.wireMockConfig;
import static java.lang.String.format;
import static java.util.Comparator.reverseOrder;
import static java.util.stream.Collectors.toList;

@Configuration
@Profile("record")
public class RecordConfiguration {

    @Bean
    public MockServers mockServers(
            final WireMockProperties properties,
            final ResponseTemplateTransformer responseTemplateTransformer
    ) {
        return new MockServers(toRecordMockServers(properties, responseTemplateTransformer));
    }

    private List<MockServer> toRecordMockServers(final WireMockProperties properties, final ResponseTemplateTransformer responseTemplateTransformer) {
        return properties.servers().stream()
                .map(serverProperties -> new RecordMockServer(
                        toWireMockServer(serverProperties, responseTemplateTransformer),
                        toRecordSpec(serverProperties)
                ))
                .collect(toList());
    }

    private WireMockServer toWireMockServer(final WireMockServerProperties properties, final ResponseTemplateTransformer responseTemplateTransformer) {
        final var rootDirectory = format("src/test/resources/wiremock/%s", properties.partner());
        createStubsDirectories(rootDirectory);

        return new WireMockServer(wireMockConfig()
                .port(properties.port())
                .withRootDirectory(rootDirectory)
                .extensions(
                        new RecordBodyTransformer(properties.pattern(), properties.replacement()),
                        responseTemplateTransformer
                )
        );
    }

    private void createStubsDirectories(final String rootDirectory) {
        final var filesDirectory = Path.of(format("%s/__files", rootDirectory));
        final var mappingsDirectory = Path.of(format("%s/mappings", rootDirectory));

        try {
            resetDirectory(filesDirectory);
            Files.createDirectories(filesDirectory);

            resetDirectory(mappingsDirectory);
            Files.createDirectories(mappingsDirectory);
        } catch (final IOException e) {
            throw new IllegalStateException(format("Unable to create stubs directory under root: %s", rootDirectory), e);
        }
    }

    private void resetDirectory(final Path directory) {
        if (Files.notExists(directory)) {
            return;
        }

        try {
            Files.walk(directory)
                    .sorted(reverseOrder())
                    .map(Path::toFile)
                    .forEach(File::delete);
        } catch (IOException e) {
            throw new IllegalStateException(format("Unable to reset directory: %s", directory), e);
        }
    }

    private RecordSpec toRecordSpec(final WireMockServerProperties properties) {
        return recordSpec()
                .forTarget(properties.endpoint())
                .ignoreRepeatRequests()
                .makeStubsPersistent(true)
                .build();
    }
}

File: proxito-wiremock/src/main/kotlin/io/fries/proxito/wiremock/record/RecordConfiguration.kt
package io.fries.proxito.wiremock.record

import com.github.tomakehurst.wiremock.WireMockServer
import com.github.tomakehurst.wiremock.client.WireMock
import com.github.tomakehurst.wiremock.core.WireMockConfiguration
import com.github.tomakehurst.wiremock.extension.Extension
import com.github.tomakehurst.wiremock.recording.RecordSpec
import io.fries.proxito.core.ApiTestContext
import io.fries.proxito.core.proxy.ProxyServer
import io.fries.proxito.core.proxy.ProxyServers
import io.fries.proxito.core.proxy.ProxyServersFactory
import io.fries.proxito.wiremock.ProxyServerProperties
import io.fries.proxito.wiremock.ROOT_DIRECTORY
import io.fries.proxito.wiremock.WireMockProperties
import io.fries.proxito.wiremock.record.transformer.ConnectionCloseResponseTransformer
import io.fries.proxito.wiremock.record.transformer.IdempotentStubIdTransformer
import io.fries.proxito.wiremock.record.transformer.RequestPatternTransformer
import io.fries.proxito.wiremock.record.transformer.ResponseDateTransformer
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.util.FileSystemUtils
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Path
import java.time.ZonedDateTime

@Configuration
class RecordConfiguration {

    @Bean
    fun proxyServersFactory(wireMockProperties: WireMockProperties, clock: () -> ZonedDateTime): ProxyServersFactory =
        ProxyServersFactory { apiTestContext ->
            ProxyServers(toRecordProxyServers(apiTestContext, wireMockProperties, clock))
        }

    private fun toRecordProxyServers(
        apiTestContext: ApiTestContext,
        wireMockProperties: WireMockProperties,
        clock: () -> ZonedDateTime
    ): List<ProxyServer> {
        return wireMockProperties.record
            .map { serverProperties ->
                RecordProxyServer(
                    toWireMockServer(apiTestContext, serverProperties, clock),
                    toRecordSpec(serverProperties)
                )
            }
    }

    private fun toWireMockServer(
        apiTestContext: ApiTestContext,
        serverProperties: ProxyServerProperties,
        clock: () -> ZonedDateTime
    ): WireMockServer {
        val rootDirectory = Path.of("$ROOT_DIRECTORY/$apiTestContext/${serverProperties.name}")
        createStubsDirectories(rootDirectory)

        return WireMockServer(
            WireMockConfiguration.wireMockConfig()
                .port(serverProperties.port)
                .withRootDirectory(rootDirectory.toString())
                .extensions(*toExtensions(serverProperties, clock))
        )
    }

    private fun toExtensions(
        serverProperties: ProxyServerProperties,
        clock: () -> ZonedDateTime
    ): Array<Extension> = listOfNotNull(
        IdempotentStubIdTransformer(serverProperties.name),
        ConnectionCloseResponseTransformer(),
        requestPatternTransformer(serverProperties),
        responseDateTransformer(serverProperties, clock)
    ).toTypedArray()

    private fun responseDateTransformer(serverProperties: ProxyServerProperties, clock: () -> ZonedDateTime) =
        serverProperties.transformers
            ?.response
            ?.let { ResponseDateTransformer(clock, it) }

    private fun requestPatternTransformer(serverProperties: ProxyServerProperties) =
        serverProperties.transformers
            ?.request
            ?.let { RequestPatternTransformer(it) }

    private fun createStubsDirectories(rootDirectory: Path) {
        try {
            FileSystemUtils.deleteRecursively(rootDirectory)
            Files.createDirectories(rootDirectory.resolve("__files"))
            Files.createDirectories(rootDirectory.resolve("mappings"))
        } catch (e: IOException) {
            throw IllegalStateException("Unable to create stubs directory under root: $rootDirectory", e)
        }
    }

    private fun toRecordSpec(serverProperties: ProxyServerProperties): RecordSpec = WireMock.recordSpec()
        .forTarget(serverProperties.endpoint)
        .ignoreRepeatRequests()
        .makeStubsPersistent(true)
        .extractTextBodiesOver(0)
        .build()
}